File: spl-net/src/main/java/bgu/spl171/net/impl/TFTP/CommandEncoders/AckCommandEncoder.java
package bgu.spl171.net.impl.TFTP.CommandEncoders;

import bgu.spl171.net.impl.TFTP.Commands.Command;
import bgu.spl171.net.impl.TFTP.Commands.Responses.Ack;
import bgu.spl171.net.impl.TFTP.EncoderDecoders.ShortEncoderDecoder;

public class AckCommandEncoder implements CommandEncoder {
    ShortEncoderDecoder shortEncoderDecoder = new ShortEncoderDecoder();
    @Override
    public byte[] encode(Command cmd) {
        Ack ack = (Ack) cmd;
        byte [] opCode =shortEncoderDecoder.encode((short)4);
        byte [] blockId = shortEncoderDecoder.encode(ack.getBlockId());
        return EncoderUtils.mergeByteArray(opCode,blockId);
    }
}

File: java to c++ encoders/AckCommandEncoder.cpp
#include "AckCommandEncoder.h"
#include "EncoderUtils.h"

namespace bgu
{
	namespace spl171
	{
		namespace net
		{
			namespace impl
			{
				namespace TFTP
				{
					namespace CommandEncoders
					{
						using bgu::spl171::net::impl::TFTP::Commands::Command;
						using bgu::spl171::net::impl::TFTP::Commands::Responses::Ack;
						using bgu::spl171::net::impl::TFTP::EncoderDecoders::ShortEncoderDecoder;

						std::vector<char> AckCommandEncoder::encode(Command *cmd)
						{
							Ack *ack = static_cast<Ack*>(cmd);
							std::vector<char> opCode = shortEncoderDecoder->encode(static_cast<short>(4));
							std::vector<char> blockId = shortEncoderDecoder->encode(ack->getBlockId());
							return EncoderUtils::mergeByteArray({opCode, blockId});
						}
					}
				}
			}
		}
	}
}
